Compiler/src/ssair/passes.jl

```diff
@@ -1511,6 +1511,10 @@ function sroa_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
             used_ssas[x.id] -= 1
         end
         ir = complete(compact)
+        # remove any use that has been optimized away by the DCE
+        for (intermediaries, defuse) in values(defuses)
+            filter!(x -> ir[SSAValue(x.idx)][:stmt] !== nothing, defuse.uses)
```

Comment:
We should probably add `x::SSAUse` annotation here.

Reply:
I don't think this is necessary, `defuse` is already typed as `SSADefUse` in the `defuses` container and `.uses` is also typed to be `SSAUse`:
```julia
struct SSADefUse
    uses::Vector{SSAUse}
    defs::Vector{Int}
end
```

Reply:
Yes, that’s true in general cases, but since `Compiler` code needs to be inferred even before inference is fully set up during bootstrap, there is a best practice of declaring method signatures as concretely as possible. 

More broadly, even in cases where inference doesn’t work perfectly, having concrete method signatures allows that the method can still be compiled, making this a generally useful technique. In this specific case, inference is very likely to succeed, but given that other parts of the code follow this convention, it might make sense to do so here as well.

That said, I’m not entirely sure if this best practice is still necessary in the current implementation of `Compiler.jl`. What do you think on this @vtjnash?

Reply:
It is mainly about readability here. A lot of untyped arguments makes it really hard to understand what is supposed to happen in the code. If only one thing can happen, then it is usually better to assert that it happened, both so that if something different happens the compiler gets an error and so the reader can see what the author expected to happen.